### DISCUSSION TITLE
Where can I find docs for `solid-js/web`'s `insert`?

### DISCUSSION THREAD

**Discussion by danieltroger** [solidjs/solid#1060] [Q&A] [answered]
I'm marrying together a codebase with a custom JSX runtime that just creates elements and solid-js with it's reactivity and JSX runtime. I want the value of a signal to be inbetween some elements created with the non-solid runtime.

I'm doing `insert(div, () => signal_containing_elements[0](), null)` now as I REd that solid does it itself and it works very nicely but since there don't seem to be any docs on `insert` I'm not completely sure if I can run into issues with solid-js deleting existing siblings? So I have

- div
-  - div already there that might replace itself with something new (might replace itself with multiple elements)
- -  value of signal

And I'm paranoid that the div already there might be deleted in some series of timing-dependent events. The [code is kinda convoluted](https://github.com/ryansolid/dom-expressions/blob/main/packages/dom-expressions/src/client.js#L363) so I thought I'd ask if that could become a problem?

Also from a really naive first look it seems like solid-js's inserting code was written before there was `Element.before`, `Element.after`, `Element.replaceWith`, `Element.append`, `Element.prepend` and `Element.replaceChildren`, a bunch of new functions that make interacting with the DOM simpler. Have you looked at looking at them or is it not worth it because bugs could be introduced? Have you benchmarked them? IIRC I benchmarked append vs appendChild once and append was faster but I might misremember 👀😅

**Comment by ryansolid** [comment, marked as answer]
This is true it does `insert` does predate a lot of those. We use `replaceWith` in some places. I was benchmarking a few of the helpers a while back and they were slower or unnoticeably different (last I checked `append` and `appendChild` were indistinguishable). I was especially interested in getting rid of that helper where it just inserts a bunch of elements (https://github.com/ryansolid/dom-expressions/blob/main/packages/dom-expressions/src/client.js#L460) but `Element.before` was worse. Probably worth looking at again as this stuff changes every few months. Especially newer things getting faster.

As for your question. Hmm.. it's possible I think. We use adjacent nodes as markers in when there are multiple dynamic children.. also when the element thinks it owns all children it will clear all children when removed. In this first case we maintain an array of all the nodes so if that changed there might be a case where it tries to re-insert the original `div`.

In general most of the `solid-js/web` helpers aren't documented as they are seen more as internal. But it does come up time to time for some them to just be useful as is.

**Reply by danieltroger** [reply]
Wow you're right, it seems like Element.before [is indeed slower](https://jsben.ch/PMrOM).

> As for your question. Hmm.. it's possible I think. We use adjacent nodes as markers in when there are multiple dynamic children.. also when the element thinks it owns all children it will clear all children when removed. In this first case we maintain an array of all the nodes so if that changed there might be a case where it tries to re-insert the original div.

Ok, thanks for the info. It works for now and I haven't hit a case where it cleared the children but I'll keep it in mind and maybe try to find another way before going into production if there's still time.
